src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(currentState, stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(currentState, stateCopy, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(currentState, stateCopy);
+        break;
+      default:
+        break;
```

Comment:
Each case in the `switch` statement should have its own logic for creating a new state object based on the action without mutating the original state or any shared object. After performing the action-specific logic, you should create a new state clone and push it to `currentState` array.